File: rep/internal/fcm/fcm_test.go
package fcm

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/json"
	"encoding/pem"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/uuid"
)

func generateServiceAccount(t *testing.T, dir string) string {
	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	pemBytes := pem.EncodeToMemory(&pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
	})

	sa := map[string]interface{}{
		"type":         "service_account",
		"project_id":   "test-project",
		"private_key":  string(pemBytes),
		"client_email": "test@test-project.iam.gserviceaccount.com",
	}
	path := filepath.Join(dir, "service-account.json")
	data, _ := json.Marshal(sa)
	if err := os.WriteFile(path, data, 0o600); err != nil {
		t.Fatalf("write service account: %v", err)
	}
	return path
}

func TestSendPushNotifications(t *testing.T) {
	dir := t.TempDir()
	serviceAccountPath := generateServiceAccount(t, dir)

	var requests []struct {
		URL  string
		Body json.RawMessage
	}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := io.ReadAll(r.Body)
		if err != nil {
			t.Fatalf("read body: %v", err)
		}
		_ = r.Body.Close()
		requests = append(requests, struct {
			URL  string
			Body json.RawMessage
		}{URL: r.URL.Path, Body: body})

		w.Header().Set("Content-Type", "application/json")
		if r.URL.Path == "/token" {
			w.Write([]byte(`{"access_token":"test-token","expires_in":3600,"token_type":"Bearer"}`))
			return
		}
		w.Write([]byte(`{"name":"projects/test/messages/1"}`))
	}))
	defer server.Close()

	envelope := Envelope{SubscriptionID: uuid.Must(uuid.NewRandom()).String()}
	pushTokens := []PushToken{
		{Provider: "fcm", Token: "fcm-token"},
	}

	result, err := SendPushNotifications(pushTokens, envelope, SendOptions{
		Service:            "test",
		ServiceAccountPath: serviceAccountPath,
		TokenURL:           server.URL + "/token",
		URL:                server.URL + "/fcm",
	})
	if err != nil {
		t.Fatalf("send: %v", err)
	}
	if result.Sent != 1 {
		t.Fatalf("expected 1 sent, got %d", result.Sent)
	}
	if len(requests) != 2 {
		t.Fatalf("expected 2 requests, got %d", len(requests))
	}

	var fcmBody map[string]interface{}
	if err := json.Unmarshal(requests[1].Body, &fcmBody); err != nil {
		t.Fatalf("parse fcm body: %v", err)
	}
	message := fcmBody["message"].(map[string]interface{})
	if message["token"] != "fcm-token" {
		t.Fatalf("token mismatch")
	}
	android := message["android"].(map[string]interface{})
	if android["priority"] != "HIGH" {
		t.Fatalf("priority mismatch")
	}
	data := message["data"].(map[string]interface{})
	if data["encryptedEnvelope"] == "" {
		t.Fatalf("missing encrypted envelope")
	}
}

File: cli-js/fcm.test.js
import assert from 'node:assert/strict';
import http from 'node:http';
import { generateKeyPairSync, randomBytes, randomUUID } from 'node:crypto';
import { mkdtemp, rm, writeFile } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { bytesToBase64Url, decryptNotification, encryptNotification } from './protocol.js';
import { sendFcmPushNotifications } from './fcm.js';

const { privateKey } = generateKeyPairSync('rsa', { modulusLength: 2048 });
const tempDir = await mkdtemp(path.join(os.tmpdir(), 'private-notify-fcm-'));
const serviceAccountPath = path.join(tempDir, 'service-account.json');
await writeFile(
  serviceAccountPath,
  JSON.stringify({
    type: 'service_account',
    project_id: 'test-project',
    private_key: privateKey.export({ type: 'pkcs8', format: 'pem' }),
    client_email: 'test@test-project.iam.gserviceaccount.com'
  }),
  'utf8'
);

const requests = [];
const server = http.createServer((request, response) => {
  let body = '';
  request.on('data', (chunk) => {
    body += chunk;
  });
  request.on('end', () => {
    requests.push({ url: request.url, body });
    response.writeHead(200, { 'content-type': 'application/json' });
    if (request.url === '/token') {
      response.end(JSON.stringify({ access_token: 'test-token', expires_in: 3600, token_type: 'Bearer' }));
    } else {
      response.end(JSON.stringify({ name: 'projects/test/messages/1' }));
    }
  });
});

await new Promise((resolve, reject) => {
  server.once('error', reject);
  server.listen(0, '127.0.0.1', resolve);
});

const { port } = server.address();
const subscription = {
  id: randomUUID(),
  name: 'fcm-test',
  key: bytesToBase64Url(randomBytes(32)),
  pushTokens: [
    {
      provider: 'fcm',
      token: 'fcm-token',
      platform: 'android',
      registeredAt: new Date().toISOString()
    }
  ],
  createdAt: new Date().toISOString()
};
const envelope = encryptNotification(subscription, {
  id: randomUUID(),
  service: 'test',
  title: 'Native title',
  body: 'Native body',
  data: {},
  createdAt: new Date().toISOString()
});

const result = await sendFcmPushNotifications(subscription.pushTokens, envelope, {
  service: 'test',
  serviceAccountPath,
  tokenUrl: `http://127.0.0.1:${port}/token`,
  url: `http://127.0.0.1:${port}/fcm`
});

assert.equal(result.sent, 1);
assert.equal(requests.length, 2);
const fcmBody = JSON.parse(requests[1].body);
assert.equal(fcmBody.message.token, 'fcm-token');
assert.equal(fcmBody.message.android.priority, 'HIGH');
assert.equal(typeof fcmBody.message.data.encryptedEnvelope, 'string');
const decrypted = decryptNotification(subscription, JSON.parse(fcmBody.message.data.encryptedEnvelope));
assert.equal(decrypted.title, 'Native title');
assert.equal(decrypted.body, 'Native body');

server.close();
await rm(tempDir, { recursive: true, force: true });
console.log('fcm ok');
